WalletWasabi/Wallets/BlockDownloadStats.cs

```diff
@@ -0,0 +1,153 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using NBitcoin.Protocol;
+using WalletWasabi.Crypto.Randomness;
+using WalletWasabi.Helpers;
+using WalletWasabi.Logging;
+
+namespace WalletWasabi.Wallets;
+
+public class BlockDownloadStats
+{
+	private double AverageLastSample { get; set; }
+	private double StandardDeviationLastSample { get; set; }
+	private int ConsecutiveFails { get; set; }
+	private IEnumerable<BlockDl> Sample { get; set; } = Enumerable.Empty<BlockDl>();
+	private List<BlockDl> History { get; } = new();
+	private object Lock { get; } = new();
+	
+	public void AddBlockDl(IPAddress from, double msElapsed, bool success)
+	{
+		lock (Lock)
+		{
+			History.Add((new BlockDl(from, msElapsed, success)));
+			Sample = History.Skip(Math.Max(0, History.Count - Const.SampleSize));
+			(AverageLastSample, StandardDeviationLastSample) = 
+				GetAverageStandardDerivation(Sample.Where(x => x.Success));
+			ConsecutiveFails = success ? 0 : ConsecutiveFails + 1;
+		}
+	}
+
+	private static (double, double) GetAverageStandardDerivation(IEnumerable<BlockDl> collection)
+	{
+		return MathUtils.AverageStandardDeviation(collection.Select(x => x.MsElapsed));
+	}
+	
+	public int NodeTimeoutStrategy(int lastTimeout)
+	{
+		lock (Lock)
+		{
+			int baseTimeout;
+			if (AverageLastSample > 0 && Sample.Count(x => x.Success) > Const.SampleMinSuccesses)
+			{
+				// Mean time spent recently to download a block including the spread
+				baseTimeout = (int)Math.Round(AverageLastSample + Math.Max(Const.MinimumSpread, StandardDeviationLastSample * Const.StandardDerivationWeight));
+			}
+			else
+			{
+				// Same but for History if we don't have enough info in sample.
+				var historyAvgSd = GetAverageStandardDerivation(History.Where(x => x.Success));
+				if (historyAvgSd.Item1 > 0)
+				{
+					baseTimeout = (int)Math.Round(historyAvgSd.Item1 + Math.Max(Const.MinimumSpread, historyAvgSd.Item2 * Const.StandardDerivationWeight));
+				}
+				else
+				{
+					// No info at all, use default value.
+					baseTimeout = Math.Max(Const.TimeoutDefaultValue, Math.Min(Const.TimeoutMaxValue, lastTimeout));
+				}
+			}
+
+			var consecutiveFailsPenalty = 0.0;
+			if (ConsecutiveFails > 0)
+			{
+				consecutiveFailsPenalty = (ConsecutiveFails - 1) * Const.BlockDlFailedPenalty *
+				                          Math.Pow(Const.BlockDlFailedMultiplier, ConsecutiveFails - 1);
+			}
+			var result = (int)Math.Round(baseTimeout + consecutiveFailsPenalty);
+			result = result > Const.TimeoutMaxValue ? Const.TimeoutMaxValue : result;
+			var timeoutDeltaLog = lastTimeout >= result ? $"(-{lastTimeout - result})" : $"(+{result - lastTimeout})";
+			Logger.LogDebug($"Timeout: {result}ms {timeoutDeltaLog} / Sample Avg: {Math.Round(AverageLastSample)}ms");
+			return result;
+		}
+	}
+
+	public bool NodeDisconnectStrategy(Node node, int connectedNodes)
```

Comment:
I am not sure about the exact mechanism here, but! Supernodes that are spying for chain analysis are having very good connections and are fast. Would your algorithm prefer those nodes in the long term? Is this strategy prefer to use the same node more times if it is fast? If yes then this could be a privacy concern. We have to carefully balance privacy and stability and speed (good ux).

Reply:
That's exactly the point of this PR.
Currently on master, nodes are disconnected every time EXCEPT (and this is really important) if we have less than 3 connected nodes.
As a result, if we found a node with good connection, we will download every block from this specific node, going too fast to let connect to other and basically never disconnecting the node.
This algo disconnects even fast node when they are used with high frecuency